### PR TITLE
get_ssh_info parameter for ec2 module

### PR DESCRIPTION
**Pull request by giacomolozito**
Hi,
this is a patch to extend the ec2 module to get SSH fingerprints and host pubkeys from the console output. This is important for security reasons and to avoid adding new SSH pubkeys to the known_hosts file manually (described in issue #575).

It works by specifying the parameter **get_ssh_info** in the ec2 module (**wait: yes** must also be present and **wait_timeout** is used to wait for ssh information). It is also possible to specify a custom regex with **get_ssh_info_regex**. Two new parameters are returned in the registered ec2 object, containing all collected fingerprints and host public keys:

``` json
                    "root_device_type": "ebs", 
                    "ssh_info_fingerprints": [
                        "1024 04:b8:80:00:94:ea:db:93:04:d0:ac:c7:3d:a4:22:6a  root@ip-172-31-33-109 (DSA)", 
                        "256 ba:09:cc:30:a7:9a:35:fd:7b:01:be:f3:5d:3b:2c:a7  root@ip-172-31-33-109 (ECDSA)", 
                        "2048 85:bf:fa:ef:68:f8:8c:98:d2:7f:f2:8d:42:39:0b:10   (RSA1)", 
                        "2048 f7:35:45:a4:f2:b1:6f:e3:85:df:ff:26:dd:10:cd:af  root@ip-172-31-33-109 (RSA)",
                    ], 
                    "ssh_info_pubkeys": [
                        "ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBJ8km+EX0vreV0koeF6Uk5MmsiAxadEwHCFJJvHkIPPRQbZovYnwwS4a/iCOU5/RlvJTMh6P7qruFbskvdkSeyc= root@ip-172-31-33-109", 
                        "2048 65537 25252192595011968799124921291908141663990399261943355348602675705850069749937630905285396766210302463668916997627523140709341653976417317771194557992723056751231726783932458978907928667539832262016396339935556847406346088261845663418364546243702535488666005793558940809882329851970841660795995578906335095712184562842943932317707574225942367281179702162389411044834227245870086380414972598220418445221732765523449108397490621464535176255881144850126284374619574858724139916831572299971152432674559727374144902374467316360612115988224292525536953325841712010390735546809624093581582576358536865167364904452376130012339", 
                        "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCukSrEYw9PatiMzFgKeDNUwf2G8zfhRSSm/3qcm986bnX99i3IAc64v8zFtV/M/jL+xZafGWS/6tvZjg/yMlD6POqIJLZjdtIpI3RKW0z0r1euZZGiaqvTOR8Mhd+56wmNjY5CEy66zV4tC/B3o3skonKyUzLBPvL5w/+SqRFhYWIn+T8bzN1GpkluoheOQOcKY845RXCsio/hlsjfgTdlYtqvq+JYmpKxljGfxICss6aY1WVSQ0RTZoxXuYgsRoAss+1Q5l23jbZdlrAU7aSKOHOHIFLMJ5rkfyEIoQIl+8lAQUrX4SAdwtnUWnnGm1+Zu5ASSjLp6dLVFxgTTJbF root@ip-172-31-33-109"
                    ], 
                    "state": "running", 
                    "state_code": 16,
```

Pros:
- gets those keys without having to blindly trust the first SSH connection
- default regex works with most Linux AMI
- regex can be customized, should the console output change in future AMI

Cons:
- console output in AWS is buffered; it takes a lot of time to become available (5 mins) thus slowing down the playbook execution when get_ssh_info is used
- RHEL 7.0/7.1 (ami-25158352/ami-05851072) are not supported, these AMI do not seem to print any key information on console output

This has been tested and confirmed to be working properly with the following Linux AMI (eu-west-1 region): Amazon 2015.03 (ami-a10897d6), Amazon 2014.09 (ami-6e7bd919), RHEL 6.5 (ami-04a00d73), SLES 12 (ami-e801af9f), Ubuntu 14.04 (ami-47a23a30).

This is an example playbook I use to auto-add RSA host keys for 3 EC2 instances:

``` yml

---
 - hosts: localhost
   connection: local
   gather_facts: False
   tasks:

    - ec2:
        key_name: foo1
        instance_type: t2.micro
        image: ami-a10897d6
        wait: yes
        wait_timeout: 600
        count: 3
        get_ssh_info: yes
        region: eu-west-1
      register: ec2_info

    # pick ssh-rsa keys among those returned for each ec2 host
    - shell: echo "{{ item.ssh_info_pubkeys | join('\n') }}" | grep -o "^ssh-rsa .*$"
      with_items:
        - "{{ ec2_info.instances }}"
      register: ec2_info_rsapub

    - known_hosts:
        key: "{{ item.item.public_ip }} {{ item.stdout }}"
        host: "{{ item.item.public_ip }}"
        state: present
      with_items:
        - "{{ ec2_info_rsapub.results }}"
```

### CODE REVIEW

**Comment by bcoca** [comment]
cc @defionscode @bennojoy

**Comment by juliedavila** [comment]
@giacomolozito apologies for the delay, can you rebase please?

**Comment by giacomolozito** [comment]
@defionscode hello, just rebased.

**Comment by bcoca** [comment]
ping @defionscode

**Comment by carmi** [comment]
I'd love to see this merged. It seems like there isn't a clear approach to solving the out-of-band ssh fingerprint verification issue when creating new cloud instances. This seems to be a solution so adding my additional +1.

**Comment by gregdek** [comment]
Thanks @giacomolozito for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

**Comment by giacomolozito** [comment]
Hello @gregdek , I have rebased the PR. This is ready_for_review

**Comment by gregdek** [comment]
Thanks @giacomolozito. @tgerla @lwade please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

**Comment by gregdek** [comment]
Thanks @giacomolozito. @tgerla @lwade please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

**Comment by robynbergeron** [comment]
@tgerla @lwade This change is still pending your review; do you have time to take a look and comment? Please comment with text 'shipit' or 'needs_revision' as appropriate.

**Comment by gregdek** [comment]
Reassigning to @ansible/core for review. Apologies @giacomolozito for the delay.

File: cloud/amazon/ec2.py [hunk 1]
```diff
@@ -253,6 +253,19 @@
       - Launch group for spot request, see U(http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/how-spot-instances-work.html#spot-launch-group)
     required: false
     default: null
+  get_ssh_info:
+    version_added: "2.0"
+    description:
+      - Used to collect ssh fingerprint information for newly-created instances, through EC2 console output. This must be used with the 'wait' parameter enabled and will usually require a wait_timeout higher than 300 seconds, as the console output is not immediately available. Parsing of the console output is done with the regular expression specified in get_ssh_info_regex.
+    required: false
+    default: null
```
Comment (juliedavila) [inline cloud/amazon/ec2.py]: The default should say `False` or `no`

**Comment by juliedavila** [comment]
First off, @giacomolozito sorry this took so long to review.

Okay so this looks good but I notice that the host keys are not returned if the instance(s) already exist.

Example 

``` yaml
---
 - hosts: localhost
   connection: local
   gather_facts: False
   tasks:

    - name: EC2 Module
      ec2:
        key_name: jdavila 
        instance_type: t2.micro
        image: ami-8fcee4e5 
        wait: yes
        wait_timeout: 600
        count_tag:
           module_testing_jdavila: 1854
        exact_count: 1
        get_ssh_info: yes
        region: us-east-1
        instance_tags:
           module_testing_jdavila: 1854
      register: ec2_info

    - debug: var=ec2_info

```

The first time this is ran, it does provide the host keys and fingerprints, subsequent runs do not provide this information.

**Comment by carmi** [comment]
Instance console output is buffered and posted shortly after instance boot, reboot, and termination. Amazon EC2 preserves the most recent 64 KB output which will be available for at least one hour after the most recent post.
I don't think the console output is a reliable method for fetching ssh fingerprints consistently. It will only work after boot, and in my experience quite slow (a few minutes to get output after box is up and running). 
I'd suggest a longer term solution use the EC2 Run Command - however this requires more config pre instance launch. 

-Evan

On Fri, Mar 11, 2016 at 12:34 PM -0800, "Jonathan Davila" notifications@github.com wrote:

First off, @giacomolozito sorry this took so long to review.

Okay so this looks good but I notice that the host keys are not returned if the instance(s) already exist.

Example 

---
- hosts: localhost
  connection: local
  gather_facts: False
  tasks:
  - name: EC2 Module
    ec2:
      key_name: jdavila 
      instance_type: t2.micro
      image: ami-8fcee4e5 
      wait: yes
      wait_timeout: 600
      count_tag:
         module_testing_jdavila: 1854
      exact_count: 1
      get_ssh_info: yes
      region: us-east-1
      instance_tags:
         module_testing_jdavila: 1854
    register: ec2_info
  - debug: var=ec2_info

The first time this is ran, it does provide the host keys and fingerprints, subsequent runs do not provide this information.

**Comment by giacomolozito** [comment]
Hi @defionscode thank you for your review.
I'll look into the required changes.

As @carmi mentioned, AWS documentation of console output implies that the buffer output might not be stored forever.

I'm thinking we could store the ssh fingerprints into AWS tags (ansible_SOMETHING) when they are first retrieved at instance boot. This would make them available and consistently retrievable by ansible in subsequent runs. Would you consider this approach acceptable?

Unfortunately, if the instance has not been provisioned with ansible or fingerprints have not been stored in instance tags in the first place, this won't work. That is due to the limitation of using console buffer ouput.

EC2 run command is something that wasn't available when this patch was originally written. I've had a quick look into it and sounds like an interesting approach but my understanding is that relies on SSM which requires an agent to be installed on the Linux instance first. That might be a bit too invasive for ansible to do just to gather ssh keys.

**Comment by juliedavila** [comment]
Considering what @carmi and @giacomolozito have brought up here is what I think
1. EC2 Run Command will not be the way we go about this as an agent is required. This is not to say I'd be against a module to manage/etc SSM, but should not be the thing that a module requires or depends on.
2. The tagging option is not acceptable as there is a hard limit to the number of tags an ec2 instance can have and it would not be a sound design to assume availability there.

The original design probably suffices, however, it needs to be documented that it will only be returned if it's a new instance.

Please update the documentation with that, update `version_added`, and update the `get_ssh_info` default to reflect the actual default of `False` or `no` Then we should be good to go I think.

**Comment by giacomolozito** [comment]
Changes done. Let me know if that's ok.

**Comment by gregdek** [comment]
Thanks @giacomolozito for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @giacomolozito for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@giacomolozito A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@giacomolozito Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@giacomolozito A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@giacomolozito Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@giacomolozito A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@giacomolozito Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@giacomolozito A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@giacomolozito Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@giacomolozito A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@giacomolozito Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@giacomolozito A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@giacomolozito Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request closed without merging**